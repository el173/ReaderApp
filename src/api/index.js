const HACKER_NEWS_BASE_URL = 'https://hacker-news.firebaseio.com/v0';

const HACKER_NEWS_TOP_STORIES = 'HACKER_NEWS_TOP_STORIES';
const HACKER_NEWS_DETAILS = 'HACKER_NEWS_DETAILS';

export {HACKER_NEWS_TOP_STORIES, HACKER_NEWS_DETAILS};

// dynamic path or query params should add as place holders to URL eg: <param1>, <param2>
export const API_LIST = [
  {
    apiName: HACKER_NEWS_TOP_STORIES,
    endpoint: `${HACKER_NEWS_BASE_URL}/topstories.json`,
    hasParams: false,
  },
  {
    apiName: HACKER_NEWS_DETAILS,
    endpoint: `${HACKER_NEWS_BASE_URL}/item/<param1>.json`,
    hasParams: true,
  },
];
