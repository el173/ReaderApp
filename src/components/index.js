import StoryCard from './StoryCard';
import CommentCard from './CommentCard';

export {StoryCard, CommentCard};
