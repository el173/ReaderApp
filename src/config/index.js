export const TOP_STORY_CHUNK_SIZE = 25;
