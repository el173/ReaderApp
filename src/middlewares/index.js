import {applyMiddleware} from 'redux';
import createSagaMiddleware from 'redux-saga';
import {createLogger} from 'redux-logger';

export const sagaMiddleware = createSagaMiddleware();

let middlewares = [sagaMiddleware];

if (__DEV__) {
  const Logger = createLogger({
    collapsed: true,
    diff: true,
  });
  middlewares = [...middlewares, Logger];
}

export const appliedMiddleware = applyMiddleware(...middlewares);
