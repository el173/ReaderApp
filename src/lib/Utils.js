import axios from 'axios';

import {API_LIST} from '../api';

/**
 *
 * @param {*} url
 * @param {*} method
 * @param {*} params
 * @returns api call response
 */
export const makeApiCall = async (url, method = 'GET', params) => {
  try {
    const configurationObject = {
      method: method,
      url: url,
    };
    if (method === 'POST') {
      // TODO:: bind params to post data
    }
    const response = await axios(configurationObject);
    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.log('AXIOS ERROR : error while making network request - ', error);
    return {
      success: false,
      info: error.code,
    };
  }
};

/**
 *
 * @param {String} apiName
 * @param  {Object} params
 * @returns relevant API endpoint
 */
export const getAPI = (apiName, params = {}) => {
  const tmpApiConfig = API_LIST.filter(
    apiConfig => apiName === apiConfig.apiName,
  )[0];
  if (tmpApiConfig.hasParams) {
    let tmpApiEndpoint = tmpApiConfig.endpoint;
    const paramKeys = Object.keys(params);
    paramKeys.map((key, index) => {
      tmpApiEndpoint = tmpApiEndpoint.replace(key, params[key]);
    });
    return tmpApiEndpoint;
  }
  return tmpApiConfig.endpoint;
};
