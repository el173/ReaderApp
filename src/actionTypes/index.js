import {GET_TOP_STORIES, GET_TOP_STORIES_OK} from './MainView';

export {GET_TOP_STORIES, GET_TOP_STORIES_OK};
