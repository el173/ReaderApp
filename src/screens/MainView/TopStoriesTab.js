import React, {useEffect, useState} from 'react';
import {connect} from 'react-redux';
import {StyleSheet, View, FlatList, ActivityIndicator} from 'react-native';

import {GET_TOP_STORIES} from '../../actionTypes';

import {StoryCard} from '../../components';

const TopStoriesTab = ({getTopStories, navigation, topStories}) => {
  const [isRefreshing, setRefreshing] = useState(false);

  useEffect(() => {
    getTopStories();
  }, []);

  useEffect(() => {
    setRefreshing(false);
  }, [topStories]);

  const onRefresh = () => {
    setRefreshing(true);
    getTopStories();
  };

  const keyExtractor = (item, index) => index;

  if (topStories.length <= 0) {
    return (
      <View style={styles.spinner}>
        <ActivityIndicator size="large" color={'#178DE7'} />
      </View>
    );
  }

  return (
    <View style={styles.container}>
      <FlatList
        data={topStories}
        onRefresh={onRefresh}
        refreshing={isRefreshing}
        keyExtractor={keyExtractor}
        renderItem={({item, index}) => (
          <StoryCard
            key={`${index}-${item.time}`}
            newsItem={item}
            navigation={navigation}
          />
        )}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  spinner: {
    position: 'absolute',
    alignSelf: 'center',
    alignItems: 'center',
    alignContent: 'center',
  },
});

const mapStateToProps = state => {
  return {
    topStories: state.mainViewReducer.topStories,
  };
};

function mapDispatchToProps(dispatch) {
  return {
    getTopStories: () => {
      dispatch({
        type: GET_TOP_STORIES,
      });
    },
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(TopStoriesTab);
