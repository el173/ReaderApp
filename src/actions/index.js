import {getTopStoriesSaga} from './MainView';

const sagas = [getTopStoriesSaga];

export const initSagas = sagaMiddleware =>
  sagas.forEach(sagaMiddleware.run.bind(sagaMiddleware));
