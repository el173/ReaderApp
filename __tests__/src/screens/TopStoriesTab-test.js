/**
 * @format
 */

import 'react-native';
import renderer from 'react-test-renderer';
import React from 'react';
import configureStore from 'redux-mock-store';
import axios from 'axios';
import {Provider} from 'react-redux';

import {makeApiCall, getAPI} from '../../../src/lib/Utils';

import TopStoriesTab from '../../../src/screens/MainView/TopStoriesTab';

import {HACKER_NEWS_TOP_STORIES} from '../../../src/api';

describe('Main View', () => {
  const mockStore = configureStore([]);

  it('renders correctly', async () => {
    async () => {
      renderer.create(<TopStoriesTab />);
    };
  });

  it('renders no with data', () => {
    const mainViewReducer = {
      topStories: [],
    };
    const store = mockStore({mainViewReducer});
    const noDataScreen = renderer
      .create(
        <Provider store={store}>
          <TopStoriesTab />
        </Provider>,
      )
      .toJSON();
    expect(noDataScreen).toMatchSnapshot();
  });

  it('renders with data', async () => {
    axios.mockResolvedValue({
      data: [
        {
          by: 'dhouston',
          descendants: 71,
          id: 8863,
          kids: [
            9224, 8917, 8952, 8958, 8884, 8887, 8869, 8873, 8940, 8908, 9005,
            9671, 9067, 9055, 8865, 8881, 8872, 8955, 10403, 8903, 8928, 9125,
            8998, 8901, 8902, 8907, 8894, 8870, 8878, 8980, 8934, 8943, 8876,
          ],
          score: 104,
          time: 1175714200,
          title: 'My YC app: Dropbox - Throw away your USB drive',
          type: 'story',
          url: 'http://www.getdropbox.com/u/2/screencast.html',
        },
      ],
    });
    const data = await makeApiCall(getAPI(HACKER_NEWS_TOP_STORIES));
    const mainViewReducer = {
      topStories: data.data,
    };
    const store = mockStore({mainViewReducer});
    const withDataScreen = renderer
      .create(
        <Provider store={store}>
          <TopStoriesTab />
        </Provider>,
      )
      .toJSON();
    expect(withDataScreen).toMatchSnapshot();
  });
});
