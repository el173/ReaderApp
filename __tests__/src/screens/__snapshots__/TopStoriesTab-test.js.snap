// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main View renders no with data 1`] = `
<View
  style={
    Object {
      "alignContent": "center",
      "alignItems": "center",
      "alignSelf": "center",
      "position": "absolute",
    }
  }
>
  <ActivityIndicator
    color="#178DE7"
    size="large"
  />
</View>
`;

exports[`Main View renders with data 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "by": "dhouston",
          "descendants": 71,
          "id": 8863,
          "kids": Array [
            9224,
            8917,
            8952,
            8958,
            8884,
            8887,
            8869,
            8873,
            8940,
            8908,
            9005,
            9671,
            9067,
            9055,
            8865,
            8881,
            8872,
            8955,
            10403,
            8903,
            8928,
            9125,
            8998,
            8901,
            8902,
            8907,
            8894,
            8870,
            8878,
            8980,
            8934,
            8943,
            8876,
          ],
          "score": 104,
          "time": 1175714200,
          "title": "My YC app: Dropbox - Throw away your USB drive",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "borderRadius": 10,
              "elevation": 3,
              "flex": 1,
              "marginBottom": 10,
              "marginHorizontal": 5,
              "padding": 15,
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 1,
                "width": 1,
              },
              "shadowOpacity": 1,
              "shadowRadius": 5,
            }
          }
        >
          <View
            style={
              Object {
                "marginBottom": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 18,
                }
              }
            >
              My YC app: Dropbox - Throw away your USB drive
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "marginBottom": 15,
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#178DE7",
                  "fontSize": 14,
                }
              }
            >
              View 33 Comment(s)
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 0.75,
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 12,
                  }
                }
              >
                Wed, 04 Apr 2007 19:16:40 GMT
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 0.25,
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 12,
                  }
                }
              >
                By: dhouston
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
